### DISCUSSION TITLE
Announcement: Script Nodes in PML version 2.3.0

### DISCUSSION THREAD

**Discussion by pml-lang** [pml-lang/pml-companion#72] [Show and tell]
Today I am thrilled to announce the biggest milestone (so far) in the evolution of PML/PDML: **_Script Nodes_**.
	
_Script nodes_ (also called _embedded scripting_) enable you to execute embedded source code when PML/PDML documents are parsed. They help to simplify and automate many aspects of the writing/building/publishing process for text and data documents.
	
To get a quick idea of what you can do with script nodes please have a look at some [examples](https://pdml-lang.github.io/docs/extensions/scripting_examples/index.html).
	
For the full documentation please refer to:

PML docs:
- chapter [Script Nodes](https://www.pml-lang.dev/docs/advanced_features/index.html#script_nodes) in [Advanced Features](https://www.pml-lang.dev/docs/advanced_features/index.html).
- [Changelog](https://www.pml-lang.dev/docs/changelog/index.html)

PDML docs:
- [Extension Nodes](https://pdml-lang.github.io/docs/extensions/user_manual/index.html#extension_nodes)
- [Reference Manual](https://pdml-lang.github.io/docs/extensions/reference_manual/index.html)
- [Scripting Examples](https://pdml-lang.github.io/docs/extensions/scripting_examples/index.html)
	
Script nodes are an experimental feature, included in the newest PML version. 
	
As always, feedback is very welcome.
	
Enjoy!

**Comment by tajmone** [comment]
This is truly amazing!

I've read through the updated PDML and PML docs and I was overwhelmed by potential of the new features. Honestly, it's something unprecedented — the sheer power handed over to end users in terms of dynamic customization of nodes and types! The potentials are limitless, and this is just the beginning.

PML will become the first DIY lightweight markup syntax, where a solid foundation allows tweaking the syntax on a per-project basis, allowing user to tailor the syntax according to needs.

I can't wait to see third party libraries being contributed, and can't even imagine what creative digital editors will come up with, but I'm sure they'll surprise us with sleek scripts and packages.

I can easily foresee how PML is going to branch into various domain specific documenting approaches, thanks to set of custom scripts, types, etc. We'll see packages dedicated to API documentation, others for scientific papers — the limit is the sky here, so anything from narrative to CYOA game books is possible.

The interplay between PDML and PML definitely adds a whole new dimension to PML as a markup syntax, where for the first time we're not bound to some third party data formats (JSON, YAML, XML) but rely instead on a format agnostic standard which is also the very foundation powering the PML engine behind the scene.

Since the first articles and PML v1, we've seen so many great changes, and how a few simple core idea managed to unfold into usable tools. Now we need to create real projects to test these features and push them to the limit, so that we can get a feel of where we're heading next in terms of what extras should be implemented natively, what as part of officially endorsed libraries, and what as third party contributions. It's now time to start creating an ecosystem around PDML and PML.

**Reply by pml-lang** [reply]
WOW! Thank you so much, Tristano. Your comment is very encouraging. I'll buy you a pizza.
